src/object.hpp

```diff
@@ -178,20 +191,323 @@ template <typename Derived> class readable_object {
+    {
+        if (obj_ == nullptr || obj_->parameterName == nullptr) {
+            [[unlikely]] return 0;
+        }
+        return static_cast<std::size_t>(obj_->parameterNameLength);
+    }
+
+    [[nodiscard]] bool empty() const noexcept { return size() == 0; }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, std::string> || std::is_same_v<T, std::string_view>
+    {
+        if (obj_ == nullptr || obj_->parameterName == nullptr) {
+            [[unlikely]] return {};
+        }
+        return {obj_->parameterName, static_cast<std::size_t>(obj_->parameterNameLength)};
+    }
+
+    template <typename T>
+    bool operator==(const T &other) const
+        requires(std::is_same_v<T, std::string_view> || std::is_same_v<T, object_key>)
+    {
+        auto s = size();
+        return s == other.size() && memcmp(data(), other.data(), s) == 0;
+    }
+
+protected:
+    const detail::object *obj_{nullptr};
+};
+
+class object_view final : public readable_object<object_view> {
+public:
+    // The default constructor results in a view without value
+    object_view() = default;
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_view(const detail::object &underlying_object) : obj_(&underlying_object) {}
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_view(const detail::object *underlying_object) : obj_(underlying_object) {}
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_view(const owned_object &ow);
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    object_view(const borrowed_object &ow);
+
+    ~object_view() = default;
+    object_view(const object_view &) = default;
+    object_view(object_view &&) = default;
+    object_view &operator=(const object_view &) = default;
+    object_view &operator=(object_view &&) = default;
+
+    [[nodiscard]] const detail::object *ptr() const noexcept { return obj_; }
+    [[nodiscard]] const detail::object &ref() const noexcept
+    {
+        assert(obj_ != nullptr);
+        return *obj_;
+    }
+
+    template <typename T> bool operator==(const T &other) const
+    {
+        if constexpr (std::is_same_v<T, std::nullptr_t>) {
+            return ptr() == nullptr;
+        } else if constexpr (std::is_same_v<std::decay_t<T>, ddwaf_object *>) {
+            return ptr() == other;
+        } else if constexpr (std::is_same_v<std::decay_t<T>, ddwaf_object>) {
+            return ptr() == &other;
+        } else if constexpr (std::is_same_v<std::decay_t<T>, object_view>) {
+            return ptr() == other.ptr();
+        } else if constexpr (std::is_same_v<std::decay_t<T>, std::string_view>) {
+            return has_value() && is_string() && as<std::string_view>() == other;
+        } else {
+            // Assume unknown types aren't equal
+            return false;
```

Comment:
This has a lot of potential to go wrong. Just static_assert that it can't happen

Reply:
Yes, you're right I noticed this as well while updating tests...